### ISSUE TITLE
EnumerateFileSystemInfos on SSD slower than HDD when Windows has the data cached

### ISSUE THREAD

**Issue by HEIC-to-JPEG-Dev** [dotnet/runtime#56898]
### Description
EnumerateFileSystemInfos against the root of an SSD that's been previously enumerated, takes longer than it does to enumerate the root of a HDD (that's been previously enumerated).

HDD
Run EnumerateFileSystemInfos at root, with RecurseSubdirectories = true. First run was 26 seconds, subsequent runs consistently <1 second.

SSD
Run EnumerateFileSystemInfos at root, with RecurseSubdirectories = true. First run was 23 seconds, subsequent runs are consistently 10 seconds.

### Configuration
EnumerationOptions = { IgnoreInaccessible = true, AttributesToSkip = FileAttributes.Device, RecurseSubdirectories = true };
foreach (var dirInfo in directoryInfo.EnumerateFileSystemInfos("*", _EnumerationOptions))...
.Net core 5 on x86 and x64 on Windows 11 x64 (same issue on Windows 10)
SSD is 4TB with 3TB in use, HDD is 4TB with 2TB in use
Memory 32GB, CPU 8 core intel series 6 with HT

### Analysis
EnumerateFileSystemInfos has an issue that might be related - you can't run this against a large mapped network drive as it will run out memory (internal memory error).

I've tried AOT, 2% performance increase.
I've tried GetFileSystemInfos - 1.5% performance increase
I've tried Release instead of debug, 5% performance increase.

However, any subsequent run against the SSD shows the SSD is idle, the processors are running at 60% and above, but performance is sub-par.
Just a guess, but is Windows not caching SSD data ?

**Comment by adamsitnik** [comment]
Hi @HEIC-to-JPEG-Dev 

> SSD is 4TB with **3**TB in use, HDD is 4TB with **2**TB in use

I don't think it's apples to apples comparision. Both disks need to contain exactly the same data. What if one of them contains fewer files? I would expect that to affect the file enumeration perf.

Moreover, from the .NET perspective, we run exactly the same code for SSD and HDD drives. 

Could you please run apples-to-apples comparision and if you confirm the problem then collect two trace files with [PerfView](https://channel9.msdn.com/Series/PerfView-Tutorial) and open a new issue in https://github.com/microsoft/Windows-Dev-Performance? 

> EnumerateFileSystemInfos has an issue that might be related - you can't run this against a large mapped network drive as it will run out memory (internal memory error)

This is something that we could defnitely take a look at. Is there any chance that you could capture a trace file with [PerfView](https://channel9.msdn.com/Series/PerfView-Tutorial) and share it with us? 

Thanks!

**Comment by HEIC-to-JPEG-Dev** [comment]
SSD is 4TB with 3TB in use, HDD is 4TB with 2TB in use

I'd need to wait until I upgrade my system to have exact duplicates of the SSD vs HDD, however, I can have both drives contain similar amounts of files and space taken. Even so, generally, HDD or second pass is always <1s but the SSD >10s. Which on a 32GB machine (assuming windows is using memory as a cache) is fishy at best.

Using EnumerateFileSystemInfos on a mapped network drive causes an internal memory issue. The numeration runs for about 25 seconds, then breaks with "System.IO.IOException,  HResult=0x800705AA, Message=Insufficient system resources exist to complete the requested service"

I've attached the PerfView output from the "run" command - please be patient if that's wrong as I've never used PerfView before.

The Perview zip file is [https://1drv.ms/u/s!AoK5hWSJoP7Hydtu7AxzrxOP-3A0RQ?e=lfPF54](url)

**Comment by HEIC-to-JPEG-Dev** [comment]
A quick update, I realised I had sensitive information on the network drive, so I moved it elsewhere. This was approximately 3.6 million files. reducing the number of files on the network share resolved the issue. So it appears to be the total number of files that was the issue.

**Comment by Jozkee** [comment]
@HEIC-to-JPEG-Dev were you able to make the apples to apples comparison?

**Comment by HEIC-to-JPEG-Dev** [comment]
Yes I ran this apples to apples. A 4tb pci 4 ssd vs a 4tb 7,200 had on local sata. I coped the same data to both, I rebooted the machine between runs and used the same code base in different scenarios. 

Results are the same, a second run on the ssd results in lower than expected performance, ssd is idle, but it can take tens of seconds to enumerate.
